### PR TITLE
Delete kernelspec entry from .ipynb file during Jovian clone

### PR DESCRIPTION
**Pull request by prabhat625**
### Summary 
Delete the kernelspec entry from the .ipynb file during jovian clone.

### CODE REVIEW

File: jovian/utils/clone.py [hunk 1]
```diff
@@ -95,12 +96,17 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
...
+            # Delete  kernalspec entry from .ipynb file 
+            if f['filename'][-6:] == ".ipynb":
+                nb_content = json.loads(get(f['rawUrl']).content.decode("utf-8"))
+                if 'metadata' in nb_content and 'kernelspec' in nb_content['metadata']: 
+                    del nb_content['metadata']['kernelspec']
+                content = bytes(json.dumps(nb_content), 'utf-8')
+            else:
+                content=get(f['rawUrl']).content
+            with open(f['filename'], 'wb') as fp: 
+                fp.write(content)
```
Comment (aakashns) [inline jovian/utils/clone.py:108]: Better move this to a helper function called `_sanitize_notebook`

File: jovian/utils/clone.py [hunk 2]
```diff
@@ -95,12 +96,17 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
...
     for f in gist['files']:
         if not f['artifact'] or include_outputs:
-            with open(f['filename'], 'wb') as fp:
-                fp.write(get(f['rawUrl']).content)
+            # Delete  kernalspec entry from .ipynb file 
+            if f['filename'][-6:] == ".ipynb":
+                nb_content = json.loads(get(f['rawUrl']).content.decode("utf-8"))
+                if 'metadata' in nb_content and 'kernelspec' in nb_content['metadata']: 
+                    del nb_content['metadata']['kernelspec']
+                content = bytes(json.dumps(nb_content), 'utf-8')
```
Comment (aakashns) [inline jovian/utils/clone.py:104]: Better move this to a helper function called `_sanitize_notebook`

File: jovian/utils/clone.py [hunk 3]
```diff
@@ -95,12 +96,17 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
...
+            else:
+                content=get(f['rawUrl']).content
+            with open(f['filename'], 'wb') as fp: 
+                fp.write(content)
 
-        # Create .jovianrc for a fresh clone
-        if fresh and f['filename'].endswith('.ipynb'):
-            set_notebook_slug(f['filename'], slug)
```
Comment (PrajwalPrashanth) [inline jovian/utils/clone.py:103]: Why is this removed ?

File: jovian/utils/clone.py [hunk 4]
```diff
@@ -95,12 +96,17 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
     log('Downloading files..')
     for f in gist['files']:
         if not f['artifact'] or include_outputs:
-            with open(f['filename'], 'wb') as fp:
-                fp.write(get(f['rawUrl']).content)
+            # Delete  kernalspec entry from .ipynb file 
+            if f['filename'][-6:] == ".ipynb":
```
Comment (PrajwalPrashanth) [inline jovian/utils/clone.py:100]: Can use `.endswith`

File: jovian/utils/clone.py (hunk 2 above)
Comment (PrajwalPrashanth) [inline jovian/utils/clone.py:104]: You can probably use `nbformat` instead of `json`

**Review by PrajwalPrashanth — CHANGES_REQUESTED** [review]

File: jovian/utils/clone.py [hunk 5]
```diff
@@ -95,12 +96,17 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
...
+                nb_content = json.loads(get(f['rawUrl']).content.decode("utf-8"))
+                if 'metadata' in nb_content and 'kernelspec' in nb_content['metadata']:
```
Comment (sonakshs) [inline jovian/utils/clone.py:102]: nit: 
```suggestion
                if nb_content.get('metadata', {}).get('kernelspec'):
```

**Comment by prabhat625** [comment]
All changes at the library are done with testing. 
@PrajwalPrashanth @sonakshs

File: jovian/utils/clone.py [hunk 6]
```diff
@@ -95,8 +96,13 @@ def clone(slug, version=None, fresh=True, include_outputs=True, overwrite=False)
     log('Downloading files..')
     for f in gist['files']:
         if not f['artifact'] or include_outputs:
-            with open(f['filename'], 'wb') as fp:
-                fp.write(get(f['rawUrl']).content)
+            # Delete  kernalspec entry from .ipynb file 
+            if f['filename'].endswith('.ipynb'):
+                content = _sanitize_notebook(f['rawUrl'])
```
Comment (sidujjain) [inline jovian/utils/clone.py:101]: you can change function definition and pass `get(f['rawUrl']).content` here

**Comment by PrajwalPrashanth** [comment]
image
I'm getting this error

**Comment by PrajwalPrashanth** [comment]
> 
> 
> image
> I'm getting this error
@prabhat625 
I tried it with a other notebook (just one cell) it does not seem to be consistent, but always consistent with this notebook. https://jovian.ml/PrajwalPrashanth/colab-improvements

I tried with `clone` without the changes of this PR, where this error would not appear for the same notebook. Can you try to clone this notebook and LMK if this persists ? Most probably some inconsistency while using `json`.

If you come across the same message for the above notebook, can you try with `nbformat` and see if that persists.

**Comment by PrajwalPrashanth** [comment]
> 
> 
> image
> I'm getting this error

This is quite flaky and not happening for all the notebooks. 
Seems to be consistent with this notebook https://jovian.ml/PrajwalPrashanth/colab-improvements/v/13 and when your env has nbformat=5.0.4 (same notebook works fine with 5.0.7, which is the default version which gets installed at the time of this comment)

File: jovian/utils/clone.py (hunk 6 above)
Comment (prabhat625) [reply jovian/utils/clone.py]: Changed.

**Review by PrajwalPrashanth — APPROVED** [review]

**Review by PrajwalPrashanth — APPROVED** [review]
